To support Spring Boot 3.0 the local deployer needs util to find open port.

cppwfs opened the pull request:
This is because SocketUtils was removed from the Spring Framework.

resolves #213

cppwfs committed:
To support Spring Boot 3.0 the local deployer needs the ability to find an open port.

This is because SocketUtils was removed from the Spring Framework.

resolves #213

onobc commented on spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/DeployerSocketUtils.java:
```diff
@@ -0,0 +1,296 @@
+ * finding available ports on {@code localhost}.
+ *
+ * <p>
+ * Within this class, a TCP port refers to a port for a {@link ServerSocket}; whereas, a
```
This looks like a lift/shift from the now removed `SocketUtils`. If so, we should probably state that somewhere in the comments. Otherwise, the copyrights/authors list raises an eyebrow. 

Also, the original reason for it being [removed from SF](https://github.com/spring-projects/spring-framework/issues/28052)

> However, these utilities make no guarantee about the subsequent availability of a given port and are therefore unreliable.

although it has been reliable for our uses so far, correct?

onobc commented on spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/DeployerSocketUtils.java:
```diff
@@ -0,0 +1,296 @@
+	 * @throws IllegalStateException if the requested number of available ports could not
+	 * be found
+	 */
+	public static SortedSet<Integer> findAvailableTcpPorts(int numRequested, int minPort, int maxPort) {
```
I am in favor of ditching all of the unused API footprint. Lets only keep what is absolutely necessary. Wdyt?

cppwfs replied:
Agreed.  I removed all unessential methods and created https://github.com/spring-cloud/spring-cloud-deployer-local/issues/215 to enhance this behavior.

onobc reviewed (changes requested):
Hi @cppwfs - thanks for getting ahead of this one. It looks good. I do have a couple requests inline though.

garyrussell commented:
FYI, this technique is unreliable - hence the removal from SF; it is possible that the port may be in use by the time you start listening on it.

The user of this needs to defend against that and try again.

cppwfs commented:
@garyrussell Agreed. This PR, was to restore functionality that was lost because of the removal of SocketUtils and restore the functionality so that affected downstream projects' functionality can be restored.  I will create an issue that suggests we use ServerSocket to identify available ports.

cppwfs committed:
Updated to remove all unessential methods.

A new issue has been opened to discuss replacing this solution with an improved solution:
https://github.com/spring-cloud/spring-cloud-deployer-local/issues/215

onobc commented on spring-cloud-deployer-local/src/main/java/org/springframework/cloud/deployer/spi/local/DeployerSocketUtils.java:
```diff
@@ -31,266 +30,148 @@
+ * of a given port and are therefore unreliable. Instead of using SocketUtils to
+ * find an available local port for a server, it is recommended that you rely on a
+ * server's ability to start on a random port that it selects or is assigned by the operating system.
+ * To interact with that server, you should query the server for the port it is currently using.
```
If we immediately deprecate this it will also nag us into the reminder that we need to replace it. Wdyt?

onobc reviewed (approved):
Thanks for the updates and creating the subsequent follow-on ticket.

cppwfs committed:
Updated with deprecation

Updated based on code review
